src/layer-surface.c

```diff
@@ -185,6 +185,7 @@ layer_surface_map (CustomShellSurface *super, struct wl_surface *wl_surface)
     zwlr_layer_surface_v1_set_exclusive_zone (self->layer_surface, self->exclusive_zone);
     layer_surface_send_set_anchor (self);
     layer_surface_send_set_margin (self);
+    zwlr_layer_surface_v1_set_exclusive_edge (self->layer_surface, self->exclusive_edge);
```

Comment:
Here and in `layer_surface_set_exclusive_edge` you should check if the compositor supports the necessary protocol version. Something like
```
uint32_t version = zwlr_layer_surface_v1_get_version (self->layer_surface);
if (version >= ZWLR_LAYER_SURFACE_V1_SET_EXCLUSIVE_EDGE_SINCE_VERSION) {
    zwlr_layer_surface_v1_set_exclusive_edge (self->layer_surface, self->exclusive_edge);
}
```
similar to the `layer_surface_set_layer`.

Now, I'm not sure what's the right fallback behavior for `layer_surface_set_exclusive_edge`. Probably anchoring to 3 edges and tweaking GTK surface size/alignment to start drawing from the requested edge.
Returning an error and leaving fallback behavior to the user could work as well.

Reply:
The library seems to be quite close to how the Wayland protocol looks, so returning an error sounds more consistent with the feel of the library.

That said, there is `gtk_layer_auto_exclusive_zone_enable` which is not present in the protocol and is handled by the library.

An automatic fallback would mean:
- For 4 edges, fall back to 3 edges and set the window size to make it appear fullscreen
- For 2 edges, like you said, 3 edges and manipulate the drawing area
- There are probably some edge cases like 0 edges and 2 opposing edges
- Exclusive edge set to one that isn't anchored to? Is that allowed?

Reply:
I just added a warning to `layer_surface_set_exclusive_edge`.